### ISSUE TITLE
Hanging issue

### ISSUE THREAD

**Issue by citrusella** [iymeko/strategitica#44]
**Describe the bug**
Tried to log into Strategitica, had it hang, let it run the script about 5 minutes (until my computer was at 7/8 of its memory in use) and then told Firefox to stop the page.

The first two log entries loaded almost immediately. I didn't see any other logs at all, until I terminated the script/page. (I actually did this a few times but decided on using this pass because the first was one I stopped fast just to see (though the log is identical) and the second was left 10 minutes but my computer was basically at max memory use and Firefox put the tab on standby when I switched tabs for something so I couldn't see the page contents anymore.

Did not get past login on any attempt.

Logs are here because the log is VERY long: https://docs.google.com/document/d/1oD97UTRIQegOjQxJBSIYaD4oR4ijS3-Zpq681Fx2ohc (and also because I can make the doc private (right now it's open with commenting enabled) if I feel like I don't want large swaths of my task list just... on the internet like that--I don't think I have anything too awful private that anyone would be looking through a near 200 page file for (or else I'd PM the file link) but...)

For reference, this is what I saw *before* termination:

```
[00:22:37]
Tags loaded successfully
[00:22:37]
Tasks loaded successfully
```

Once the script was stopped, in addition to the log fleshing itself out, the user header loaded in behind the login window (i.e. with my name/level/class and XP/HP/MP). But the login window hung around and no tasks appeared.

**Expected behavior**

Well, I *wanted* to log in (though I didn't expect to given I'd been having the issue) but I was actually surprised that the log itself seemed to hang until the script was dead. Somehow I thought it was going to be informative for what it did rather than potentially what it didn't. XP

**Desktop (please complete the following information):**
 - OS: Windows 10 version 2004
 - Browser: Firefox
 - Version: 78.10.0 ESR

**Comment by rugbyopera** [comment]
I have essentially identical experience as citrusella.

OS: Windows 10 Home version 1909
Browser1: Chrome 90.0.4430.93 (Official Build) (64-bit)
Browser2: Brave Version 1.23.73 Chromium: 90.0.4430.85 (Official Build) (64-bit) (clean install, no extensions)
Browser3: Firefox 88.0

Logs from Chrome console:
Before submitting UUID/API key:

Error with Permissions-Policy header: Unrecognized feature: 'interest-cohort'.
Navigated to https://iymeko.github.io/strategitica/?logs=true
iymeko.github.io/:1 [DOM] Input elements should have autocomplete attributes (suggested: "current-password"): (More info: https://goo.gl/9p2vKq) <input type=​"password" class=​"form-control" name=​"apiToken" id=​"api-token">​
[Violation] Forced reflow while executing JavaScript took 41ms
onloadwff.js:71 [Violation] 'setTimeout' handler took 100ms

Immediately after clicking "Let's Go!" to submit UUID/key:
VM11:53 XHR finished loading: GET "https://habitica.com/api/v3/tags?_=1619619970167".
XMLHttpRequest.send @ VM11:53
(anonymous) @ VM17:1
send @ jquery.js:10099
ajax @ jquery.js:9682
(anonymous) @ user.js:42
create @ user.js:77
loadAll @ index.js:40
(anonymous) @ index.js:30
dispatch @ jquery.js:5429
elemData.handle @ jquery.js:5233

VM11:53 XHR finished loading: GET "https://habitica.com/api/v3/tags?_=1619619970167".
XMLHttpRequest.send @ VM11:53
(anonymous) @ VM17:1
send @ jquery.js:10099
ajax @ jquery.js:9682
(anonymous) @ user.js:42
create @ user.js:77
loadAll @ index.js:40
(anonymous) @ index.js:30
dispatch @ jquery.js:5429
elemData.handle @ jquery.js:5233
VM11:53 XHR finished loading: GET "https://habitica.com/api/v3/tasks/user?_=1619619970168".

Once the page finally timed out:
VM11:53 XHR finished loading: GET "https://habitica.com/api/v3/tags?_=1619619970167".
XMLHttpRequest.send @ VM11:53
(anonymous) @ VM17:1
send @ jquery.js:10099
ajax @ jquery.js:9682
(anonymous) @ user.js:42
create @ user.js:77
loadAll @ index.js:40
(anonymous) @ index.js:30
dispatch @ jquery.js:5429
elemData.handle @ jquery.js:5233
VM11:53 XHR finished loading: GET "https://habitica.com/api/v3/tasks/user?_=1619619970168".

Like citrusella, the log didn't fill in until the page timed out.
![screenshot](https://user-images.githubusercontent.com/82522888/116423033-4f659d00-a80e-11eb-9890-c38e4470ecf4.jpg)

**Comment by iymeko** [comment]
Didn't mean to close, although I am hoping #45 will fix this... @citrusella and @rugbyopera, let me know if you see any improvement?

**Comment by iymeko** [comment]
Also, @citrusella:

> because I can make the doc private (right now it's open with commenting enabled) if I feel like I don't want large swaths of my task list just... on the internet like that--I don't think I have anything too awful private that anyone would be looking through a near 200 page file for (or else I'd PM the file link) but...)

I get what you're saying about having your personal info publicly available. I'm thinking of showing task IDs in the logs instead of the task name, and maybe a checkbox to switch between showing task IDs vs task titles. And probably something similar for your username/display name (i.e. just show your user ID?). Would that be enough or do you think more is necessary?

**Comment by rugbyopera** [comment]
With ?logs=true, it's behaving identically today as it did yesterday (except I see the new Loading bar). Same i=359 and everything.

With ?days=1 (no logs=true), I'm getting a couple new errors in the console.
Access to XMLHttpRequest at 'https://habitica.com/api/v3/user?_=1619718823779' from origin 'https://iymeko.github.io' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.
VM29:53 GET https://habitica.com/api/v3/user?_=1619718823779 net::ERR_FAILED
XMLHttpRequest.send @ VM29:53
(anonymous) @ VM35:1
send @ jquery.js:10099
ajax @ jquery.js:9682
getUserInfo @ user.js:145
create @ user.js:38
loadAll @ index.js:46
(anonymous) @ index.js:32
dispatch @ jquery.js:5429
elemData.handle @ jquery.js:5233
jquery.js:4046 jQuery.Deferred exception: Cannot use 'in' operator to search for 'message' in undefined TypeError: Cannot use 'in' operator to search for 'message' in undefined
    at Object.<anonymous> (https://iymeko.github.io/strategitica/main.bundle.js:35112:39)
    at mightThrow (https://iymeko.github.io/strategitica/main.bundle.js:4557:29)
    at process (https://iymeko.github.io/strategitica/main.bundle.js:4625:12) undefined
jQuery.Deferred.exceptionHook @ jquery.js:4046
process @ jquery.js:3834
setTimeout (async)
(anonymous) @ jquery.js:3868
fire @ jquery.js:3496
fireWith @ jquery.js:3626
fire @ jquery.js:3634
fire @ jquery.js:3496
fireWith @ jquery.js:3626
deferred.<computed> @ jquery.js:3964
fire @ jquery.js:3496
fireWith @ jquery.js:3626
done @ jquery.js:9788
(anonymous) @ jquery.js:10039
error (async)
send @ jquery.js:10067
ajax @ jquery.js:9682
getUserInfo @ user.js:145
create @ user.js:38
loadAll @ index.js:46
(anonymous) @ index.js:32
dispatch @ jquery.js:5429
elemData.handle @ jquery.js:5233

**Comment by rugbyopera** [comment]
> Also, @citrusella:
> 
> > because I can make the doc private (right now it's open with commenting enabled) if I feel like I don't want large swaths of my task list just... on the internet like that--I don't think I have anything too awful private that anyone would be looking through a near 200 page file for (or else I'd PM the file link) but...)
> 
> I get what you're saying about having your personal info publicly available. I'm thinking of showing task IDs in the logs instead of the task name, and maybe a checkbox to switch between showing task IDs vs task titles. And probably something similar for your username/display name (i.e. just show your user ID?). Would that be enough or do you think more is necessary?

Task IDs are fine to share. Nobody can do anything with the task ID without being logged in as the user, and anyone logged in as the user already has access.

**Comment by iymeko** [comment]
@citrusella @rugbyopera - how about now...?

**Comment by rugbyopera** [comment]
I have just a minute, so brief report today. Still hanging. I can now see the (blank) calendar load. Ten minutes later, it hangs with the out of memory message. The loop starting on line 32 of helpers.js is where it crashed, with i=359.

Interestingly, after it crashed and the log filled in ("Data loaded successfully!" and "User info loaded successfully. Here's some info about you"), I noticed the timestamps were from 8 minutes earlier than they became visible on my screen.

**Comment by iymeko** [comment]
@rugbyopera, do you have Log XMLHttpRequests checked in your Chrome console settings? And do you have verbose checked in the console levels dropdown thing?

**Comment by rugbyopera** [comment]
@iymeko Yes to both.

**Comment by iymeko** [comment]
@rugbyopera well s***. If you're not seeing anything telling in your console then to be honest I'm at a loss. I only have a couple other ideas I could think of to try which are to

1. Fix the couple of warnings I see in the console although I doubt they're part of the problem.
2. Make a test account and basically do a bunch of things to try to get it to hang, cuz it just does not hang for me.
3. Replace jquery with React. I think it would be a big improvement but based on your description of the issue I'm iffy on whether it would help you at all. Also I'm a beginner in React so it'd take a long time probably.

If you have any better suggestions it would be much appreciated. Sorry.

**Comment by iymeko** [comment]
@rugbyopera @citrusella success! I think anyway. Basically it seems it was an issue with data types that I didn't handle properly. Really sorry about that. Please let me know if it's still hanging for you.

**Comment by citrusella** [comment]
I hadn't been able to test (or to put it more accurately: I wasn't able to test again until about two or three days ago and *then* I was subjecting my computer to heavy stress and didn't want to put more on its plate *or* have that affect any testing it out on my end), but I just tried it and it's success indeed!

**Comment by rugbyopera** [comment]
Working perfectly for me today, on both my phone and PC. Thanks so much. As a visual person, Strategitica helps me organize my Habitica setup.

**Comment by iymeko** [comment]
Awesome, so glad it's working for you both again.

> Hmm, it has decided one of my never-due tasks is due on Saturdays, though. 0_o

That is weird... So it's a daily without a due date? I'll make an issue for this. Hopefully I can reproduce this.
